storage: fix decommissioning of absent node

tbg opened the pull request:
When a node is decommissioned in absentia, the gateway node may not be
aware of its most recent liveness entry. Before this bug fix, the
gateway would fail the decommissioning process.  Instead, it now falls
back to reading the liveness record from the KV store.

Fixes #17995.

Release note (bug fix): Decommissioning a node that has already been
terminated now works in all cases. Success previously depended on
whether the gateway node "remembered" the absent decommissionee.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/23082)
<!-- Reviewable:end -->

nvanbenschoten commented:
Review status: 0 of 2 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---

*[pkg/storage/node_liveness.go, line 276 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23082#-L6HksaOCC3QFrUsHckb:-L6HksaOCC3QFrUsHckc:b-tqxjtd) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5cd0c3eb98f0ee040668a05a13668478aca92c2/pkg/storage/node_liveness.go#L276)):*
> ```Go
> 		// results in unintentional no-ops (due to the Gossip lag); this could be
> 		// observed by users in principle, too.
> 		dummyLiveness := &Liveness{NodeID: nodeID} // always catches condition failed
> ```

Why don't we create a new method to retrieve the liveness instead of this obscure approach of sending a cput and hoping for the condition failed error? I don't think we need the retry logic in `updateLiveness` if we're only performing a read.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23082)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23082#-:-L6IA0MKw5inmwe356G8:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
TFTRs!

---

Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion.

---

*[pkg/storage/node_liveness.go, line 276 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23082#-L6HksaOCC3QFrUsHckb:-L6IH5ZXnedWjrP3cqo7:bfbq99i) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5cd0c3eb98f0ee040668a05a13668478aca92c2/pkg/storage/node_liveness.go#L276)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Why don't we create a new method to retrieve the liveness instead of this obscure approach of sending a cput and hoping for the condition failed error? I don't think we need the retry logic in `updateLiveness` if we're only performing a read.
</blockquote></details>

Good point, done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23082#-:-L6IH4Th9w2K5QEPuDyp:bz5mair)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[pkg/storage/node_liveness.go, line 276 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23082#-L6IO7uQbLo2M0ZiviEU:-L6IO7uQbLo2M0ZiviEV:b-9r1p8g) ([raw file](https://github.com/cockroachdb/cockroach/blob/4381f4eb46ca76feee68c1cfc7cc1196842e72ee/pkg/storage/node_liveness.go#L276)):*
> ```Go
> 		// results in unintentional no-ops (due to the Gossip lag); this could be
> 		// observed by users in principle, too.
> 		oldLiveness := &Liveness{}
> ```

Nit: put `oldLiveness` on the stack instead of dereferencing everywhere below.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23082#-:-L6IOW4oDc4epvgeoXgG:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tbg committed:
storage: fix decommissioning of absent node

When a node is decommissioned in absentia, the gateway node may not be
aware of its most recent liveness entry. Before this bug fix, the
gateway would fail the decommissioning process. Instead, we now read the
liveness record from the KV store in this path.

Fixes #17995.

Release note (bug fix): Decommissioning a node that has already been
terminated now works in all cases. Success previously depended on
whether the gateway node "remembered" the absent decommissionee.

tschottdorf commented:
Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/storage/node_liveness.go, line 276 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23082#-L6IO7uQbLo2M0ZiviEU:-L6IgItqL6351MPo01uE:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4381f4eb46ca76feee68c1cfc7cc1196842e72ee/pkg/storage/node_liveness.go#L276)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Nit: put `oldLiveness` on the stack instead of dereferencing everywhere below.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23082)*
<!-- Sent from Reviewable.io -->
